File: web/public/estimated-strike-time.js
/**
 * Ocenjeni čas udara: Europe/Ljubljana, nato zaokrožitev na 5 min (samo prikaz).
 * Enaka logika kot strele_archive/estimated_strike_time.py
 */
(function (global) {
  "use strict";

  var LJ = "Europe/Ljubljana";

  function parseInstant(input) {
    if (input == null || input === "") return null;
    if (input instanceof Date) {
      return isNaN(input.getTime()) ? null : input;
    }
    var s = String(input);
    var d = new Date(s.indexOf("T") >= 0 ? s : s + "T12:00:00");
    return isNaN(d.getTime()) ? null : d;
  }

  function ljParts(instant) {
    var parts = new Intl.DateTimeFormat("en-GB", {
      timeZone: LJ,
      year: "numeric",
      month: "2-digit",
      day: "2-digit",
      hour: "2-digit",
      minute: "2-digit",
      second: "2-digit",
      hourCycle: "h23",
    }).formatToParts(instant);
    function n(type) {
      for (var i = 0; i < parts.length; i++) {
        if (parts[i].type === type) return Number(parts[i].value);
      }
      return NaN;
    }
    return {
      year: n("year"),
      month: n("month"),
      day: n("day"),
      hour: n("hour"),
      minute: n("minute"),
      second: n("second"),
    };
  }

  function ljWallToUtc(year, month, day, hour, minute) {
    var utc = new Date(Date.UTC(year, month - 1, day, hour, minute, 0));
    for (var i = 0; i < 4; i++) {
      var p = ljParts(utc);
      var wanted = Date.UTC(year, month - 1, day, hour, minute);
      var actual = Date.UTC(p.year, p.month - 1, p.day, p.hour, p.minute);
      var diff = wanted - actual;
      if (diff === 0) break;
      utc = new Date(utc.getTime() + diff);
    }
    return utc;
  }

  function pad2(n) {
    return String(n).padStart(2, "0");
  }

  /** UTC Date zaokroženega lokalnega časa, ali null. */
  function roundEstimatedStrikeInstant(input) {
    var d = parseInstant(input);
    if (!d) return null;
    var p = ljParts(d);
    var totalMin = p.hour * 60 + p.minute + p.second / 60;
    var rounded = Math.round(totalMin / 5) * 5;
    var year = p.year;
    var month = p.month;
    var day = p.day;
    if (rounded >= 24 * 60) {
      rounded -= 24 * 60;
      var next = new Date(Date.UTC(year, month - 1, day + 1));
      year = next.getUTCFullYear();
      month = next.getUTCMonth() + 1;
      day = next.getUTCDate();
    } else if (rounded < 0) {
      rounded += 24 * 60;
      var prev = new Date(Date.UTC(year, month - 1, day - 1));
      year = prev.getUTCFullYear();
      month = prev.getUTCMonth() + 1;
      day = prev.getUTCDate();
    }
    return ljWallToUtc(year, month, day, Math.floor(rounded / 60), rounded % 60);
  }

  function formatEstimatedStrikeTime(input) {
    var rounded = roundEstimatedStrikeInstant(input);
    if (!rounded) return "—";
    var p = ljParts(rounded);
    return pad2(p.hour) + "." + pad2(p.minute);
  }

  function formatEstimatedStrikeDateTime(input) {
    var rounded = roundEstimatedStrikeInstant(input);
    if (!rounded) return input ? String(input) : "—";
    var p = ljParts(rounded);
    return p.day + ". " + p.month + ". " + p.year + ", " + pad2(p.hour) + "." + pad2(p.minute);
  }

  global.StrelkoEstimatedStrikeTime = {
    roundEstimatedStrikeInstant: roundEstimatedStrikeInstant,
    formatEstimatedStrikeTime: formatEstimatedStrikeTime,
    formatEstimatedStrikeDateTime: formatEstimatedStrikeDateTime,
  };
})(typeof window !== "undefined" ? window : globalThis);

File: strele_archive/estimated_strike_time.py
"""Ocenjeni čas udara: Europe/Ljubljana, nato zaokrožitev na 5 min (samo prikaz)."""

from __future__ import annotations

from datetime import datetime, timedelta, timezone
from zoneinfo import ZoneInfo

LJUBLJANA_TZ = ZoneInfo("Europe/Ljubljana")

ESTIMATED_STRIKE_TIME_NOTE = (
    "Časi posameznih udarov so ocenjeni na podlagi razpoložljivih podatkov "
    "in zaokroženi na 5 minut."
)


def _to_aware_utc(ts: datetime) -> datetime:
    if ts.tzinfo is None:
        return ts.replace(tzinfo=timezone.utc)
    return ts.astimezone(timezone.utc)


def round_estimated_strike_datetime(ts: datetime) -> datetime:
    """
    Pretvori v Europe/Ljubljana, zaokroži na najbližjih 5 minut.
    Vrne lokalni datetime (sekunde = 0). Ne spreminja shranjenih podatkov.
    """
    local = _to_aware_utc(ts).astimezone(LJUBLJANA_TZ)
    total_min = (
        local.hour * 60
        + local.minute
        + local.second / 60
        + local.microsecond / 60_000_000
    )
    rounded = int(round(total_min / 5) * 5)
    day = local.date()
    if rounded >= 24 * 60:
        rounded -= 24 * 60
        day = day + timedelta(days=1)
    elif rounded < 0:
        rounded += 24 * 60
        day = day - timedelta(days=1)
    hour, minute = divmod(rounded, 60)
    return datetime(day.year, day.month, day.day, hour, minute, 0, tzinfo=LJUBLJANA_TZ)


def format_estimated_strike_time(ts: datetime | None) -> str:
    if ts is None:
        return "—"
    local = round_estimated_strike_datetime(ts)
    return f"{local.hour:02d}.{local.minute:02d}"


def format_estimated_strike_datetime(ts: datetime | None) -> str:
    """Prikaz kot v widgetu: ``19. 7. 2026, 12.15``."""
    if ts is None:
        return "—"
    local = round_estimated_strike_datetime(ts)
    return f"{local.day}. {local.month}. {local.year}, {local.hour:02d}.{local.minute:02d}"
